sxhkd.1
.TH "sxhkd" 1 "June 23, 2013" "Sxhkd"
.SH NAME
sxhkd - Simple X HotKey Daemon
.SH SYNOPSIS
.PP
\f[B]sxhkd\f[] [\f[I]OPTIONS\f[]] [\f[I]EXTRA_CONFIG\f[] ...]
.SH DESCRIPTION
.PP
sxhkd is a simple X hotkey daemon with a powerful and compact
configuration syntax.
.SH OPTIONS
.TP
.B \f[C]\-h\f[]
Print the synopsis to standard output and exit.
.RS
.RE
.TP
.B \f[C]\-v\f[]
Print the version information to standard output and exit.
.RS
.RE
.TP
.B \f[C]\-t\f[] \f[I]TIMEOUT\f[]
Timeout in seconds for the recording of chord chains.
.RS
.RE
.TP
.B \f[C]\-c\f[] \f[I]CONFIG_FILE\f[]
Read the main configuration from the given file.
.RS
.RE
.TP
.B \f[C]\-r\f[] \f[I]REDIR_FILE\f[]
Redirect the commands output to the given file.
.RS
.RE
.SH CONFIGURATION
.PP
Each line of the configuration file is interpreted as so:
.IP \[bu] 2
If it is empty or starts with \f[C]#\f[], it is ignored.
.IP \[bu] 2
If it starts with a space, it is read as a command.
.IP \[bu] 2
Otherwise, it is read as a hotkey.
.PP
General syntax:
.IP
.nf
\f[C]
HOTKEY
\ \ \ \ COMMAND

HOTKEY\ :=\ CHORD_1\ ;\ CHORD_2\ ;\ \&...\ ;\ CHORD_n
CHORD_i\ :=\ [MODIFIER_i][\@|!|:]KEYSYM_i
MODIFIER_i\ :=\ MODIFIER_i1\ +\ MODIFIER_i2\ +\ \&...\ +\ MODIFIER_ik
\f[]
.fi
.PP
The valid modifier names are: \f[I]super\f[], \f[I]hyper\f[],
\f[I]meta\f[], \f[I]alt\f[], \f[I]control\f[], \f[I]ctrl\f[],
\f[I]shift\f[], \f[I]mode_switch\f[], \f[I]lock\f[], \f[I]mod1\f[],
\f[I]mod2\f[], \f[I]mod3\f[], \f[I]mod4\f[] and \f[I]mod5\f[].
.PP
The keysym names are given by the output of \f[B]xev\f[].
.PP
Hotkeys and commands can be spread across multiple lines by ending each
partial line with a backslash character.
.PP
When multiple chords are separated by semicolons, the hotkey is a chord
chain: the command will only be executed after receiving each chord of
the chain in consecutive order.
.PP
If \f[B]\@\f[] is added at the beginning of the keysym, the command will
be run on key release events, otherwise on key press events.
.PP
If \f[B]!\f[] is added at the beginning of the keysym, the command will
be run on motion notify events and must contain two integer conversion
specifications which will be replaced by the \f[I]x\f[] and \f[I]y\f[]
coordinates of the pointer relative to the root window referential (the
only valid button keysyms for this type of hotkeys are:
\f[I]button1\f[], \&..., \f[I]button5\f[]).
.PP
If \f[B]:\f[] is added at the beginning of the keysym, the captured
event will be replayed for the other clients.
.PP
Mouse hotkeys can be defined by using one of the following special
keysym names: \f[I]button1\f[], \f[I]button2\f[], \f[I]button3\f[],
\&..., \f[I]button24\f[].
.PP
The hotkey and the command may contain sequences of the form
{\f[I]STRING_1\f[],\f[I]\&...\f[],\f[I]STRING_N\f[]}.
.PP
In addition, the sequences can contain ranges of the form
\f[I]A\f[]\-\f[I]Z\f[] where \f[I]A\f[] and \f[I]Z\f[] are alphanumeric
characters.
.PP
The underscore character represents an empty sequence element.
.PP
What is actually executed is \f[B]SHELL\f[] \f[I]\-c\f[]
\f[B]COMMAND\f[], which means you can use environment variables in
\f[B]COMMAND\f[].
.PP
\f[B]SHELL\f[] will be the content of the first defined environment
variable in the following list: \f[B]SXHKD_SHELL\f[], \f[B]SHELL\f[].
.PP
If \f[B]sxhkd\f[] receives a \f[I]SIGUSR1\f[] signal, it will reload its
configuration file.
.PP
If no configuration file is specified via the \f[I]\-c\f[] option, the
following is used: \f[B]$XDG_CONFIG_HOME/sxhkd/sxhkdrc\f[].
.SH AUTHOR
.MT baskerville@lavabit.com
Bastien Dejean
.ME
.SH LINKS
.UR https://github.com/baskerville/sxhkd
Homepage
.UE


sxhkd.c
#include <xcb/xcb_keysyms.h>
#include <xcb/xcb_event.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/select.h>
#include <fcntl.h>
#include <ctype.h>
#include <getopt.h>
#include <signal.h>
#include <stdbool.h>
#include "helpers.h"
#include "types.h"
#include "parse.h"
#include "grab.h"
#include "sxhkd.h"

void hold(int sig)
{
    if (sig == SIGHUP || sig == SIGINT || sig == SIGTERM)
        running = false;
    else if (sig == SIGUSR1)
        reload = true;
    else if (sig == SIGALRM)
        bell = true;
}

void setup(void)
{
    dpy = xcb_connect(NULL, NULL);
    if (xcb_connection_has_error(dpy))
        err("Can't open display.\n");
    xcb_screen_t *screen = xcb_setup_roots_iterator(xcb_get_setup(dpy)).data;
    if (screen == NULL)
        err("Can't acquire screen.\n");
    root = screen->root;
    if ((shell = getenv(SXHKD_SHELL_ENV)) == NULL && (shell = getenv(SHELL_ENV)) == NULL)
        err("The '%s' environment variable is not defined.\n", SHELL_ENV);
    symbols = xcb_key_symbols_alloc(dpy);
    hotkeys = hotkeys_tail = NULL;
}

void cleanup(void)
{
    PUTS("cleanup");
    hotkey_t *hk = hotkeys;
    while (hk != NULL) {
        hotkey_t *tmp = hk->next;
        destroy_chain(hk->chain);
        free(hk);
        hk = tmp;
    }
    hotkeys = hotkeys_tail = NULL;
}

void reload_cmd(void)
{
    PUTS("reload");
    cleanup();
    load_config(config_file);
    for (int i = 0; i < num_extra_confs; i++)
        load_config(extra_confs[i]);
    ungrab();
    grab();
}

void key_button_event(xcb_generic_event_t *evt, uint8_t event_type)
{
    xcb_keysym_t keysym = XCB_NO_SYMBOL;
    xcb_button_t button = XCB_NONE;
    bool replay_event = false;
    uint16_t modfield = 0;
    uint16_t lockfield = num_lock | caps_lock | scroll_lock;
    parse_event(evt, event_type, &keysym, &button, &modfield);
    modfield &= ~lockfield & MOD_STATE_FIELD;
    if (keysym != XCB_NO_SYMBOL || button != XCB_NONE) {
        hotkey_t *hk = find_hotkey(keysym, button, modfield, event_type, &replay_event);
        if (hk != NULL) {
            run(hk->command);
        }
    }
    switch (event_type) {
        case XCB_BUTTON_PRESS:
        case XCB_BUTTON_RELEASE:
            if (replay_event)
                xcb_allow_events(dpy, XCB_ALLOW_REPLAY_POINTER, XCB_CURRENT_TIME);
            else
                xcb_allow_events(dpy, XCB_ALLOW_SYNC_POINTER, XCB_CURRENT_TIME);
            break;
        case XCB_KEY_PRESS:
        case XCB_KEY_RELEASE:
            if (replay_event)
                xcb_allow_events(dpy, XCB_ALLOW_REPLAY_KEYBOARD, XCB_CURRENT_TIME);
            else
                xcb_allow_events(dpy, XCB_ALLOW_SYNC_KEYBOARD, XCB_CURRENT_TIME);
            break;
    }
    xcb_flush(dpy);
}

void motion_notify(xcb_generic_event_t *evt, uint8_t event_type)
{
    xcb_motion_notify_event_t *e = (xcb_motion_notify_event_t *) evt;
    /* PRINTF("motion notify %X %X %u\n", e->child, e->detail, e->state); */
    uint16_t lockfield = num_lock | caps_lock | scroll_lock;
    uint16_t buttonfield = e->state >> 8;
    uint16_t modfield = e->state & ~lockfield & MOD_STATE_FIELD;
    xcb_button_t button = 1;
    while (~buttonfield & 1 && button < 5) {
        buttonfield = buttonfield >> 1;
        button++;
    }
    hotkey_t *hk = find_hotkey(XCB_NO_SYMBOL, button, modfield, event_type, NULL);
    if (hk != NULL) {
        char command[MAXLEN];
        snprintf(command, sizeof(command), hk->command, e->root_x, e->root_y);
        run(command);
    }
}

int main(int argc, char *argv[])
{
    char opt;
    config_path = NULL;
    timeout = TIMEOUT;

    while ((opt = getopt(argc, argv, "vht:c:r:")) != -1) {
        switch (opt) {
            case 'v':
                printf("%s\n", VERSION);
                exit(EXIT_SUCCESS);
                break;
            case 'h':
                printf("sxhkd [-h|-v|-t TIMEOUT|-c CONFIG_FILE|-r REDIR_FILE] [EXTRA_CONFIG ...]\n");
                exit(EXIT_SUCCESS);
                break;
            case 't':
                timeout = atoi(optarg);
                break;
            case 'r':
                redir_fd = open(optarg, O_WRONLY | O_CREAT, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
                if (redir_fd == -1)
                    warn("Failed to open the command redirection file.\n");
                break;
            case 'c':
                config_path = optarg;
                break;
        }
    }

    num_extra_confs = argc - optind;
    extra_confs = argv + optind;

    if (config_path == NULL) {
        char *config_home = getenv(CONFIG_HOME_ENV);
        if (config_home != NULL)
            snprintf(config_file, sizeof(config_file), "%s/%s", config_home, CONFIG_PATH);
        else
            snprintf(config_file, sizeof(config_file), "%s/%s/%s", getenv("HOME"), ".config", CONFIG_PATH);
    } else {
        strncpy(config_file, config_path, sizeof(config_file));
    }

    signal(SIGINT, hold);
    signal(SIGHUP, hold);
    signal(SIGTERM, hold);
    signal(SIGUSR1, hold);
    signal(SIGALRM, hold);

    setup();
    get_standard_keysyms();
    get_lock_fields();
    load_config(config_file);
    for (int i = 0; i < num_extra_confs; i++)
        load_config(extra_confs[i]);
    grab();

    xcb_generic_event_t *evt;
    int fd = xcb_get_file_descriptor(dpy);

    fd_set descriptors;

    reload = bell = chained = false;
    running = true;

    xcb_flush(dpy);

    while (running) {
        FD_ZERO(&descriptors);
        FD_SET(fd, &descriptors);

        if (select(fd + 1, &descriptors, NULL, NULL, NULL) > 0) {
            while ((evt = xcb_poll_for_event(dpy)) != NULL) {
                uint8_t event_type = XCB_EVENT_RESPONSE_TYPE(evt);
                switch (event_type) {
                    case XCB_KEY_PRESS:
                    case XCB_KEY_RELEASE:
                    case XCB_BUTTON_PRESS:
                    case XCB_BUTTON_RELEASE:
                        key_button_event(evt, event_type);
                        break;
                    case XCB_MOTION_NOTIFY:
                        motion_notify(evt, event_type);
                        break;
                    default:
                        PRINTF("received event %u\n", event_type);
                        break;
                }
                free(evt);
            }
        }

        if (reload) {
            signal(SIGUSR1, hold);
            reload_cmd();
            reload = false;
        }

        if (bell) {
            signal(SIGALRM, hold);
            abort_chain();
            bell = false;
        }

        if (xcb_connection_has_error(dpy)) {
            warn("The server closed the connection.\n");
            running = false;
        }
    }

    if (redir_fd != -1)
        close(redir_fd);
    ungrab();
    cleanup();
    xcb_key_symbols_free(symbols);
    xcb_disconnect(dpy);
    return EXIT_SUCCESS;
}
